libp2p/Host.cpp

```diff
@@ -60,930 +60,926 @@ ReputationManager::ReputationManager()
-				toConnect.push_back(p.second);
-		}
-	}
-
-	for (auto p: toConnect)
-		if (p->peerType == PeerType::Required && reqConn++ < m_idealPeerCount)
-			connect(p);
-	
-	if (!m_netPrefs.pin)
-	{
-		unsigned pendingCount = 0;
-		DEV_GUARDED(x_pendingNodeConns)
-			pendingCount = m_pendingPeerConns.size();
-		int openSlots = m_idealPeerCount - peerCount() - pendingCount + reqConn;
-		if (openSlots > 0)
-			for (auto p: toConnect)
-				if (p->peerType == PeerType::Optional && openSlots--)
-					connect(p);
-	}
-
-	auto runcb = [this](boost::system::error_code const& error) { run(error); };
-	m_timer->expires_from_now(boost::posix_time::milliseconds(c_timerInterval));
-	m_timer->async_wait(runcb);
+    if (!m_run)
+    {
+        // reset NodeTable
+        DEV_GUARDED(x_nodeTable)
+            m_nodeTable.reset();
+
+        // stopping io service allows running manual network operations for shutdown
+        // and also stops blocking worker thread, allowing worker thread to exit
+        m_ioService.stop();
+
+        // resetting timer signals network that nothing else can be scheduled to run
+        DEV_GUARDED(x_runTimer)
+            m_timer.reset();
+
+        m_timerReset.notify_all();
+        return;
+    }
+
+    if (auto nodeTable = this->nodeTable()) // This again requires x_nodeTable, which is why an additional variable nodeTable is used.
+        nodeTable->processEvents();
+
+    // cleanup zombies
+    DEV_GUARDED(x_connecting)
+        m_connecting.remove_if([](std::weak_ptr<RLPXHandshake> h){ return h.expired(); });
+    DEV_GUARDED(x_timers)
+        m_timers.remove_if([](std::shared_ptr<boost::asio::deadline_timer> t)
+        {
+            return t->expires_from_now().total_milliseconds() < 0;
+        });
+
+    keepAlivePeers();
+    
+    // At this time peers will be disconnected based on natural TCP timeout.
+    // disconnectLatePeers needs to be updated for the assumption that Session
+    // is always live and to ensure reputation and fallback timers are properly
+    // updated. // disconnectLatePeers();
+
+    // todo: update peerSlotsAvailable()
+    
+    list<shared_ptr<Peer>> toConnect;
+    unsigned reqConn = 0;
+    {
+        RecursiveGuard l(x_sessions);
+        for (auto const& p: m_peers)
+        {
+            bool haveSession = havePeerSession(p.second->id);
+            bool required = p.second->peerType == PeerType::Required;
+            if (haveSession && required)
+                reqConn++;
+            else if (!haveSession && p.second->shouldReconnect() && (!m_netPrefs.pin || required))
+                toConnect.push_back(p.second);
+        }
+    }
+
+    for (auto p: toConnect)
+        if (p->peerType == PeerType::Required && reqConn++ < m_idealPeerCount)
+            connect(p);
+    
+    if (!m_netPrefs.pin)
+    {
+        int openSlots = m_idealPeerCount - peerCount() - m_pendingPeerConns.size() + reqConn;
```

Comment:
Are there any type conversions happening here? What are the chances of overflow?

Reply:
Hmm, all the values on the right of assignment are `unsigned`
Not sure now if it's  safe...
realistically none of these values go anywhere near `max<int>()`

Reply:
I will better rewrite it without cast to `int`